### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by arikmaor**
### NOT TESTED YET

Unit tests will be added in a future PR
Manual tests will be done when integrating with `lycan` (creating the `build` command)

### CODE REVIEW

File: src/_handler.ts [hunk 1]
```diff
@@ -1,13 +1,118 @@
...
 app.set('trust proxy', true);
 
-app.get('*', async (_req, res) => {
-  res.setHeader('Content-Type', 'application/json');
-  res.end(JSON.stringify({
-    hello: 'world'
-  }));
+app.use(express.json());
+
+app.get('/build', async (req, res) => {
```
Comment (vogre) [inline src/_handler.ts]: `app.post`?

File: src/_handler.ts [hunk 2]
```diff
@@ -1,13 +1,118 @@
...
+  }
+
+  const { getUrl, putUrl, endLabel } = params;
+
+  try {
+    const downloadDir = await getSourceCode(getUrl);
+    const buildDir = await build(downloadDir);
+    const tarballPath = await createTarball(buildDir);
+    const digest = await uploadTarFile(tarballPath, putUrl);
+    res.sendStatus(200);
```
Comment (vogre) [inline src/_handler.ts]: `res.status(200)` if there's a following `res.end` I think.

File: src/_handler.ts [hunk 3]
```diff
@@ -1,13 +1,118 @@
...
+  } catch (err) {
+    console.error(err);
+    res.sendStatus(500);
+  } finally {
+    console.log(endLabel);
+  }
 });
 
+function validateRequest(req: Request): ReqParams {
+  const { getUrl, putUrl, endLabel } = req.body
```
Comment (vogre) [inline src/_handler.ts]: Lint is missing for `;`.

File: src/_handler.ts [hunk 4]
```diff
@@ -1,13 +1,118 @@
...
+
+  console.log('Download complete!');
+
+  return downloadDir;
+}
+
+async function uploadTarFile(tarballPath: string, url: string): Promise<string> {
+  const fileContent = await readFile(tarballPath);
+  const digest = createHash('sha256')
+    .update(fileContent)
```
Comment (vogre) [inline src/_handler.ts]: Might be worth streaming here, as our functions by default run with 256MiB only, I tried approach like [this](https://github.com/binaris/riab/blob/2e010987c42ca0aaad3c19c6937f193447b04f78/src/main.ts#L234).

**Review by vogre — COMMENTED** [review]
Thanks!

File: src/_handler.ts (hunk 2 above)
Comment (arikmaor) [reply src/_handler.ts]: 10x

File: src/_handler.ts (hunk 1 above)
Comment (arikmaor) [reply src/_handler.ts]: 10x
fixed

File: src/_handler.ts (hunk 4 above)
Comment (arikmaor) [reply src/_handler.ts]: 10x!
Fixed

File: src/_handler.ts [hunk 5]
```diff
@@ -1,13 +1,120 @@
...
+      cb();
+    })),
+  });
+
+  if (!ok) {
+    throw new Error(`Upload failed with status: ${status} - ${statusText}`);
+  }
+  console.log('done!');
+
+  return hash.toString();
```
Comment (vogre) [inline src/_handler.ts]: `toString('hex')` to keep the format.

File: src/_handler.ts [hunk 6]
```diff
@@ -1,13 +1,120 @@
-import express from 'express';
+import path from 'path';
+import { tmpdir } from 'os';
+import { createHash } from 'crypto';
+import through from 'through2';
+import express, { Request } from 'express';
+import { createReadStream, readFile, mkdtemp } from 'mz/fs';
```
Comment (vogre) [inline src/_handler.ts]: In `node12` runtime possible `import { promises as fs } from 'fs';`

**Review by vogre — COMMENTED** [review]
Cool, thanks!

File: src/_handler.ts (hunk 6 above)
Comment (arikmaor) [reply src/_handler.ts]: Ack

File: src/_handler.ts (hunk 5 above)
Comment (arikmaor) [reply src/_handler.ts]: 10x, fixed

File: package.json [hunk 7]
```diff
@@ -6,18 +6,27 @@
...
   },
   "license": "UNLICENSED",
   "dependencies": {
+    "@reshuffle/build-utils": "^0.1.0",
     "@reshuffle/code-transform": "^0.1.0",
     "@reshuffle/local-proxy": "^0.3.5",
     "@reshuffle/server-function": "^0.2.2",
-    "@reshuffle/utils-subprocess": "0.0.2",
-    "express": "^4.17.1"
+    "is-plain-object": "^3.0.0",
```
Comment (vogre) [inline package.json]: 3x more popular than `lodash.isplainobject`!

File: builder-function.Dockerfile.j2 [hunk 8]
```diff
@@ -16,5 +16,6 @@ COPY Makefile .
 
 # and the app itself
 COPY tsconfig.json ./
```
Comment (vogre) [inline builder-function.Dockerfile.j2]: Nit: L18-L20 Can be `COPY tsconfig.json src .eslintrc.js ./`?

**Review by vogre — APPROVED** [review]
Thanks!

This is not tested yet but the skeleton for development is in place so LGTM!

File: src/_handler.ts [hunk 9]
```diff
@@ -1,13 +1,131 @@
...
+  let params: ReqParams;
+  try {
+    params = validateRequest(req);
+  } catch (err) {
+    console.error(err);
+    res.sendStatus(400);
+    return;
+  }
+
+  // TODO: destruct 'env' here and provide it to 'build()' in a few lines once build-utils is updated.
```
[the remark this answers is not in the document]
Comment (arikmaor) [reply src/_handler.ts]: Makes more sense for me to add this an a parameter to `build`
I'm not sure what kind of issues updating `process.env` will bring

File: builder-function.Dockerfile.j2 (hunk 8 above)
Comment (arikmaor) [reply builder-function.Dockerfile.j2]: Will check and create a new PR if possible

**Pull request merged**